Bottom: 347c56960bbaf0d9a20bdf30720a78f5e6679ca4
Top:    baa6878d49d4ff6b6106d4715f88b8049729f8c0
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-23 17:26:29 +0800

vhost_net: passing raw xdp buff to tun

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 554cd57..b25ae56 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -2394,6 +2394,28 @@ static void tun_sock_write_space(struct sock *sk)
 	kill_fasync(&tfile->fasync, SIGIO, POLL_OUT);
 }
 
+static void tun_xdp_one(struct tun_struct *tun,
+			struct tun_file *tfile,
+			struct xdp_buff *xdp)
+{
+	struct bpf_prog *xdp_prog;
+	int buflen;
+
+	preempt_disable();
+	rcu_read_lock();
+
+	xdp_prog = rcu_dereference(tun->xdp_prog);
+	if (xdp_prog) {
+		xdp_set_data_meta_invalid(xdp);
+		xdp->rxq = &tfile->xdp_rxq;
+		buflen = *(int *)xdp->data_hard_start;
+		tun_do_xdp(tun, tfile, xdp_prog, xdp, buflen);
+	}
+
+	rcu_read_unlock();
+	preempt_enable();
+}
+
 static int tun_sendmsg(struct socket *sock, struct msghdr *m, size_t total_len)
 {
 	int ret;
@@ -2403,9 +2425,16 @@ static int tun_sendmsg(struct socket *sock, struct msghdr *m, size_t total_len)
 	if (!tun)
 		return -EBADFD;
 
+	if (m->msg_control) {
+		struct xdp_buff *xdp = m->msg_control;
+		tun_xdp_one(tun, tfile, xdp);
+		goto out;
+	}
+
 	ret = tun_get_user(tun, tfile, m->msg_control, &m->msg_iter,
 			   m->msg_flags & MSG_DONTWAIT,
 			   m->msg_flags & MSG_MORE);
+out:
 	tun_put(tun);
 	return ret;
 }
diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 1245c85..03fe8f2 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -529,6 +529,7 @@ static void handle_tx(struct vhost_net *net)
 {
 	struct vhost_net_virtqueue *nvq = &net->vqs[VHOST_NET_VQ_TX];
 	struct vhost_virtqueue *vq = &nvq->vq;
+	struct xdp_buff xdp;
 	unsigned out, in;
 	int head;
 	struct msghdr msg = {
@@ -625,6 +626,10 @@ static void handle_tx(struct vhost_net *net)
 			ubufs = NULL;
 		}
 
+		err = vhost_net_build_xdp(nvq, &msg.msg_iter, &xdp);
+		if (!err) {
+			msg.msg_control = &xdp;
+		}
 		total_len += len;
 		if (total_len < VHOST_NET_WEIGHT &&
 		    !vhost_vq_avail_empty(&net->dev, vq) &&
